common/src/stack/checklist/checklist.py

```diff
@@ -0,0 +1,466 @@
+				b.installKernel  = o['kernel']
+				b.installRamdisk = o['ramdisk']
+				b.installArgs    = o['args']
+
+			key = b.osaction + '-os'
+			if key in bootactionMap:
+				o = bootactionMap[key]
+				b.osKernel  = o['kernel']
+				b.osRamdisk = o['ramdisk']
+				b.osArgs    = o['args']
+
+	#
+	# Gather all information relevant to monitor Backend installation
+	# and store it in a dictionary for ease of access
+	#
+	def getBackendInfo(self):
+		hnameBackendMap = self.getHosts()
+		self.getHostAttr(hnameBackendMap)
+		ipBackendMap = self.getHostInterfaces(hnameBackendMap)
+
+		# Build a bootaction dictionary for ease of access
+		bootactionMap = self.getBootactions()
+
+		self.populateBootActionInfo(ipBackendMap.values(), bootactionMap)
+		self.ipBackendMap = ipBackendMap
+
+	# Find Backend object based on MAC address
+	def getBackendByMac(self, mac):
+		backendList = self.ipBackendMap.values()
+		for b in backendList:
+			if mac in b.macList:
+				return b
+		return None
+
+	# Process TFTP messages based on internal state information
+	def processTftp(self, sm):
+		pxeFile = sm.msg.strip()
+		backend = self.ipBackendMap[sm.ipAddr]
+		retVal = Checklist.SUCCESS_TFTP
+
+		# Check if TFTP message components match backend details
+		if '/' in pxeFile:
+			pxeArr = pxeFile.split('/')
+			hexip = pxeArr[1]
+			backendIpArr = backend.ipList[0].split('.')
+			backendHexIp = '{:02X}{:02X}{:02X}{:02X}' \
+				.format(*map(int, backendIpArr))
+
+			#
+			# Ignore TFTP pxelinux.cfg messages in the post
+			# install phase
+			#
+			if (backend.isPostPkgInstallStage() or 
+				backendHexIp != pxeArr[-1]):
+				retVal = Checklist.IGNORE_TFTP_MSG
+		elif pxeFile == backend.installKernel:
+			sm.state = State.VMLinuz_RRQ_Install
+		elif pxeFile == backend.installRamdisk:
+			sm.state = State.Initrd_RRQ
+		elif pxeFile in backend.osKernel:
+			sm.state = State.Rebooting_HDD
+		else:
+			retVal = Checklist.REFRESH_STATE
+
+		return retVal
+
+	#
+	# Look through DHCP State arr and restore messages that arrived
+	# within 60 seconds before TFTP messages.
+	#
+	def restoreDhcpMsgs(self, sm):
+		backend = self.ipBackendMap[sm.ipAddr]
+		clearFlag = False
+		index = -1
+
+		# Ignore all DHCP messages in the post installation phase
+		if backend.isPostPkgInstallStage():
+			return
+
+		for s in backend.dhcpStateArr:
+			d = int(sm.time - s.time)
+
+			if 0 <= d < 5:
+				backend.stateArr.append(s)
+				clearFlag = True
+				if index == -1:
+					index = len(backend.stateArr)
+
+		if clearFlag:
+			if index > 0 and backend.stateArr[index-2].state not in  \
+				[State.DHCPDISCOVER, State.DHCPOFFER, \
+				State.DHCPACK, State.DHCPREQUEST]:
+				# Delete state messages from previous install
+				del backend.stateArr[0:index]
+			backend.dhcpStateArr.clear()
+
+	#
+	# Process messages from the shared Queue
+	#
+	def processQueueMsgs(self):
+		while not self.shutdownFlag.is_set():
+			sm = self.queue.get()
+			self.log.debug(sm)
+			with self.lock:
```

Comment:
Locks cannot be avoided since there is a need to share Backend installation state information (list of StateMessage objects received for a backend) between Checklist main thread and the CheckTimeout thread. Its not for the database fetches. Even in the CheckTimeout thread, it does a deepcopy of the Backend state information and looks for timeout based on the deepcopy inorder to avoid contention between Checklist main thread and CheckTimeout. This will result in some consistency issues for timeout messages but I felt we can deal with that.

Reply:
The only way to get around this is to write the Backend StateMessage information to redis so the CheckTimeout can read this info from the database instead of sharing state info with Checklist main. What do you think?

Reply:
Using Redis as the distributed state store would probably be cleaner. For timeouts (I haven't looked at the code deeply), couldn't you just store an BACKENDIP_state = (LAST_STATE, EPOCH_ADDED) and then look up that key and if EPOCH_ADDED is too old then you know you timed out? Seems like you just need the last successful state and not the list of all states.